src/middlewared/middlewared/plugins/directoryservices_/ipa_join_mixin.py

```diff
@@ -49,14 +49,71 @@ class IPAJoinMixin:
 
     def _ipa_leave(self, job: Job, ds_type: DSType, domain: str):
         """
-        This method is currently left unimplemented because it is not
-        currently exposed in public APIs, but will be as part of general
-        directory services redesign in future TrueNAS release.
+        Leave the IPA domain
         """
-        raise CallError(
-            'Functionality to automatically leave an IPA domain has not yet '
-            'been implemented'
-        )
+        ldap_config = self.middleware.call_sync('ldap.config')
+        ipa_config = self.middleware.call_sync('ldap.ipa_config', ldap_config)
+
+        if ipa_config['domain'] != domain:
+            raise CallError(f'{domain}: TrueNAS is joined to {ipa_config["domain"]}')
+
+        job.set_progress(0, 'Deleting NFS and SMB service principals.')
+        self._ipa_del_spn()
+
+        job.set_progress(10, 'Removing DNS entries.')
+        self.unregister_dns(ipa_config['host'], False)
+
+        # now leave IPA
+        job.set_progress(30, 'Leaving IPA domain.')
+        try:
+            join = subprocess.run([
+                IPACTL, '-a', IpaOperation.LEAVE.name
+            ], check=False, capture_output=True)
+            _parse_ipa_response(join)
+        except Exception:
+            self.logger.warning(
+                'Failed to disable TrueNAS machine account in IPA domain '
+                'further action by the IPA administrator to fully remove '
```

Comment:
This warning doesn't read right

Reply:
Yeah, danger of writing a warning over the weekend. I'll english-ify it.